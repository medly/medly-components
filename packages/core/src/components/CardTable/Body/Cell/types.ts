import { ColumnConfig, Data } from '../../types';

export interface Props {
    data: any;
    column: ColumnConfig;
    rowId: any;
    isRowClickDisabled?: boolean;
    rowData: Data;
}

export interface StyledProps {
    align?: 'left' | 'center' | 'right';
    wrapText?: boolean;
}
