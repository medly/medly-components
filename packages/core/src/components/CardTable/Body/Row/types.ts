import { WithThemeProp } from '@medly-components/utils';

export interface Props extends WithThemeProp {
    onClick?: () => void;
    isRowClickDisabled?: boolean;
    withWhiteBackground?: boolean;
    gridTemplateColumns: string;
}
