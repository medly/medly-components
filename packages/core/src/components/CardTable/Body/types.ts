import { ColumnConfig, Data } from '../types';

export interface Props {
    data: Data[];
    columns: ColumnConfig[];
    onRowClick?: (rowData: Record<string, unknown>) => void;
    rowClickDisableKey?: string;
    uniqueKeyName?: string;
    withWhiteBackground?: boolean;
}
