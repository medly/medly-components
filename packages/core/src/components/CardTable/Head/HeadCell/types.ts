import { ColumnConfig, SortOrder } from '../../types';

export interface Props {
    defaultSortOrder?: SortOrder | 'none';
    column: ColumnConfig;
    sortField?: string;
    withWhiteBackground?: boolean;
    onSortChange?: (field: string, order: SortOrder) => void;
}

export interface StyledProps {
    isActive?: boolean;
    sortable?: boolean;
    withWhiteBackground?: boolean;
    align?: 'left' | 'center' | 'right';
}
