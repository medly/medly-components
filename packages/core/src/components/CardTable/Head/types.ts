import { ColumnConfig, SortOrder } from '../types';

export interface Props {
    defaultSortField?: string;
    defaultSortOrder?: SortOrder;
    withWhiteBackground?: boolean;
    columns: ColumnConfig[];
    onSort?: (field: string, order: SortOrder) => void;
}
