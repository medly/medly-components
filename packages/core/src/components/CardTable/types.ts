import { HTMLProps } from '@medly-components/utils';
import type { FC } from 'react';

export type SortOrder = 'asc' | 'desc';

export interface Data {
    [key: string]: any;
}

export interface ColumnConfig {
    /** Title of the column */
    title: string;
    /** Field name in the data */
    field: string;
    /** Set it true to show sort icons */
    sortable?: boolean;
    /** Column content alignment */
    align?: 'left' | 'right' | 'center';
    /** Column Text content wrapping */
    wrapText?: boolean;
    /** Provide the fraction of the of the total width of the row */
    fraction?: number;
    /** Custom component */
    component?: FC<{ data: any; rowId?: any; disabled?: boolean; rowData?: Data }>;
}
export interface CardTableProps extends Omit<HTMLProps<HTMLOListElement>, 'data' | 'type'> {
    /** Table data */
    data: Array<{
        [key: string]: any;
    }>;
    /** Column configuration */
    columns: ColumnConfig[];
    /** Key name in data to be used as unique id */
    uniqueKeyName?: string;
    /** Default Sort Field*/
    defaultSortField?: string;
    /** Default Sort Order*/
    defaultSortOrder?: SortOrder;
    /** Function to be called on click of sort icon */
    onSort?: (field: string, order: SortOrder) => void;
    /** Function to be called on row click */
    onRowClick?: (rowData: Record<string, unknown>) => void;
    /** Key name to disable row click */
    rowClickDisableKey?: string;
    /** Key name to select background color */
    withWhiteBackground?: boolean;
}
