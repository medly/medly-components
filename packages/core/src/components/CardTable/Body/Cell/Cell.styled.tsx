import styled, { css } from 'styled-components';
import Text from '../../../Text';
import { StyledProps } from './types';

const wrapTextStyle = () => css`
    ${Text.Style} {
        text-overflow: ellipsis;
        overflow: hidden;
        white-space: nowrap;
    }
`;

export const Cell = styled('td')<StyledProps>`
    padding: 0;
    padding: 1.5rem 2.4rem;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: ${({ align }) => (align === 'left' ? 'flex-start' : align === 'right' ? 'flex-end' : 'center')};
    overflow: hidden;

    ${({ wrapText }) => wrapText && wrapTextStyle()}
`;

Cell.defaultProps = {
    align: 'left',
    wrapText: true
};
