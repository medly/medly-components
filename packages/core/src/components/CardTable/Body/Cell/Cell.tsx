import React, { useMemo } from 'react';
import Text from '../../../Text';
import * as Style from './Cell.styled';
import { Props } from './types';

export const Cell: React.FC<Props> = React.memo(props => {
    const { data, rowId, column, isRowClickDisabled, rowData } = props,
        { component: Component, align, wrapText } = column;

    const FormattedCell = useMemo(
        () => (Component ? <Component {...{ data, rowId, rowData, disabled: isRowClickDisabled }} /> : <Text title={data}>{data}</Text>),
        [data, rowId, isRowClickDisabled, Component]
    );

    return (
        <Style.Cell wrapText={wrapText} align={align}>
            {FormattedCell}
        </Style.Cell>
    );
});
Cell.displayName = 'Cell';
