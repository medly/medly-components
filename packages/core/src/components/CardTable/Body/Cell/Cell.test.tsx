import { render } from '@testing-library/react';
import React from 'react';
import Text from '../../../Text';
import { ColumnConfig } from '../../types';
import { Cell } from './Cell';
import { StyledProps } from './types';

const renderer = (columnConfig: ColumnConfig) =>
    render(
        <table>
            <tbody>
                <tr>
                    <Cell column={columnConfig} data="text" rowId="1" rowData={{}} />
                </tr>
            </tbody>
        </table>
    );

describe('CardTable Cell', () => {
    it('should render cell with text properly', () => {
        const { container } = renderer({
            title: 'Column 1',
            field: 'column1'
        });
        expect(container).toMatchSnapshot();
    });

    test.each([false, true])('should render cell with text component with wrap text as %s', wrapText => {
        const config: ColumnConfig = {
                wrapText,
                title: 'Column 1',
                field: 'column1',
                component: Text
            },
            { container } = renderer(config);
        expect(container).toMatchSnapshot();
    });

    const positions: Required<StyledProps>['align'][] = ['left', 'center', 'right'];

    test.each(positions)('should be able to %p align ', align => {
        const { container } = renderer({
            align,
            title: 'Column 1',
            field: 'column1'
        });
        expect(container).toMatchSnapshot();
    });
});
