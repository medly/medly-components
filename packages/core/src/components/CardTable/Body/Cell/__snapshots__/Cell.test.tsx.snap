// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardTable Cell should be able to "center" align 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  padding: 0;
  padding: 1.5rem 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: hidden;
}

.c0 .c1 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

<div>
  <table>
    <tbody>
      <tr>
        <td
          class="c0"
        >
          <span
            class="c1 c2"
            title="text"
          >
            text
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CardTable Cell should be able to "left" align 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  padding: 0;
  padding: 1.5rem 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
}

.c0 .c1 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

<div>
  <table>
    <tbody>
      <tr>
        <td
          class="c0"
        >
          <span
            class="c1 c2"
            title="text"
          >
            text
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CardTable Cell should be able to "right" align 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  padding: 0;
  padding: 1.5rem 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  overflow: hidden;
}

.c0 .c1 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

<div>
  <table>
    <tbody>
      <tr>
        <td
          class="c0"
        >
          <span
            class="c1 c2"
            title="text"
          >
            text
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CardTable Cell should render cell with text component with wrap text as false 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  padding: 0;
  padding: 1.5rem 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
}

<div>
  <table>
    <tbody>
      <tr>
        <td
          class="c0"
        >
          <span
            class="c1"
            data="text"
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CardTable Cell should render cell with text component with wrap text as true 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  padding: 0;
  padding: 1.5rem 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
}

.c0 .c1 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

<div>
  <table>
    <tbody>
      <tr>
        <td
          class="c0"
        >
          <span
            class="c1 c2"
            data="text"
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CardTable Cell should render cell with text properly 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  padding: 0;
  padding: 1.5rem 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
}

.c0 .c1 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

<div>
  <table>
    <tbody>
      <tr>
        <td
          class="c0"
        >
          <span
            class="c1 c2"
            title="text"
          >
            text
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
